src/python/pants/backend/python/util_rules/interpreter_constraints.py

```diff
@@ -233,8 +233,71 @@ def requires_python38_or_newer(self) -> bool:
             allowed_versions=["3.8", "3.9", "3.10"], prior_version="3.7"
         )
 
+    def partition_by_major_minor_versions(self) -> tuple[InterpreterConstraints, ...]:
+        """Create a distinct InterpreterConstraints value for each CPython major-minor version that
+        is compatible with the original constraints."""
+        if any(req.project_name != "CPython" for req in self):
+            raise AssertionError(
+                "This function only works with CPython interpreter constraints for now."
+            )
```

Comment:
I'm not sure what to do here if you're using PyPy. Probably we should teach this function how to handle that too?

Reply:
It feels like in cases where we can't explode the range like this, we could enforce that users are using equality match(es) instead, so the function becomes a noop?

Reply:
Perhaps - I'm in general confused how PyPy constraints work? I tried looking into PyPy yesterday and realized there is both the PyPy version _and_ their Python implementation. Like PyPy 7.3.5 can be either PyPy3.7 or PyPy2.7.

If you express the constraint as `PyPy==2.7.*`, then problem solved. If you express as `PyPy==7.3.5`, idk what to do. I'll ask John in development if he knows.

Reply:
> If you express as `PyPy==7.3.5`, idk what to do. I'll ask John in development if he knows.

I was thinking that we could basically just validate that any non-CPython interpreter req:
1. Uses equality (`==`)
2. Includes the major/minor versions at least.

...which doesn't seem like too bad a limitation.